### PR TITLE
Feature #156538237 - Add colouring by metadata

### PR DESCRIPTION
**Pull request by monicajianu**
- Extended `TsnePlotServiceDao` to retrieve t-SNE plot points without an associated k.
- Extended `TsnePlotService` to add metadata values to a set of t-SNE plot points, given a specific metadata category (such as inferred cell type).
- Split the `CellMetadataService` class to make a new Dao class, responsible for Solr queries.
- Refactored metadata retrieval related classes.
- Added new `/json/experiments/{experimentAccession}/tsneplot/{perplexity}/metadata/{metadata}` endpoint to retrive t-SNE points grouped by metadata values.
- Changed React router to include `colouredBy` and `metadata` URL query parameters.
- Added more complex logic to rewrite the URL query parameters, based on the type of colouring that has been chosen.
- Testing.

TODO: Update the package version for `expression-atlas-experiment-page-tsne-plot` once the [companion pull request](https://github.com/ebi-gene-expression-group/scxa-experiment-page-tsne-plot-view/pull/9) is approved and a new package published.

### CODE REVIEW

File: sc/src/main/javascript/bundles/experiment-page/src/TSnePlotViewRoute.js [hunk 1]
```diff
@@ -27,15 +31,32 @@ const TSnePlotViewRoute = (props) => {
...
                     selectedPerplexity={Number(search.perplexity) || props.perplexities[Math.round((perplexities.length - 1) / 2)]}
                     geneId={search.geneId || ``}
                     height={800}
                     onSelectGeneId={ (geneId) => { updateUrlSearch({ name: `geneId`, value: geneId }) } }
-                    onChangeK={ (k) => { updateUrlSearch({ name: `k`, value: k }) } }
-                    onChangePerplexity={ (perplexity) => { updateUrlSearch({ name: `perplexity`, value: perplexity }) } }
+                    onChangePerplexity={ (perplexity) => { updateUrlSearch({ name: `perplexity`, value: perplexity }) }}
+                    onChangeColourBy={
+                      (colourByCategory, colourByValue) => {
+                        const query = new URLSearchParams(history.location.search)
```
Comment (alfonsomunozpomer) [inline sc/src/main/javascript/bundles/experiment-page/src/TSnePlotViewRoute.js]: Interesting! I’ve been using URI.js for such a long time that I have no idea of the state of official APIs. I’ll have a look at it a bit more in-depth.

File: base/src/main/java/uk/ac/ebi/atlas/solr/cloud/fullanalytics/SingleCellAnalyticsCollectionProxy.java [hunk 2]
```diff
@@ -52,8 +52,14 @@ public SingleCellAnalyticsCollectionProxy(SolrClient solrClient) {
...
 
-    public QueryResponse query(SolrQueryBuilder<SingleCellAnalyticsCollectionProxy> solrQueryBuilder) {
-        return query(solrQueryBuilder.build());
+    public static String metadataFieldNameToDisplayName(String metadataFieldName) {
+        // The metadata fields are either characteristics or factors
+        if (metadataFieldName.startsWith("characteristic")) {
+            return characteristicFieldNameToDisplayName(metadataFieldName);
+        }
+        else {
+            return factorFieldNameToDisplayName(metadataFieldName);
```
Comment (alfonsomunozpomer) [inline base/src/main/java/uk/ac/ebi/atlas/solr/cloud/fullanalytics/SingleCellAnalyticsCollectionProxy.java]: Probably both `factorFieldNameToDisplayName` and `characteristicFieldNameToDisplayName` could be refactored to something like `metadataFieldToDisplayName(String fieldName, String prefix)` (where `prefix` would be `factor_`, `characteristic_` or an empty string?

File: sc/src/test/java/uk/ac/ebi/atlas/metadata/CellMetadataDaoIT.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        String experimentAccession = jdbcUtils.fetchRandomSingleCellExperimentAccession();
+        String cellId = "FOOBAR";
+
+        List<SingleCellAnalyticsCollectionProxy.SingleCellAnalyticsSchemaField> metadataFieldNames = subject.getMetadataFieldNames(experimentAccession);
+
+        assertThat(metadataFieldNames).isNotEmpty();
+
+        metadataFieldNames.forEach(field -> {
+            assertThat(subject.getMetadataValueForCellId(experimentAccession, field, cellId)).isNotPresent();
+        });
```
Comment (alfonsomunozpomer) [inline sc/src/test/java/uk/ac/ebi/atlas/metadata/CellMetadataDaoIT.java]: In this test and the a few ones below, you can chain both assertions with [`allMatch` or `allSatisfy`](http://joel-costigliola.github.io/assertj/core-8/api/index.html):
```
assertThat(metadataFieldNames)
    .isNotEmpty()
    // Predicate that returns a boolean
    .allMatch(field -> subject.getMetadataValueForCellId(experimentAccession, field, cellId).isEmpty());
```

Alternatively:
```
assertThat(metadataFieldNames)
    .isNotEmpty()
    // Consumer that makes assertions
    .allSatisfy(field -> assertThat(subject.getMetadataValueForCellId(experimentAccession, field, cellId)).isNotPresent());
```

File: sc/src/test/java/uk/ac/ebi/atlas/metadata/CellMetadataDaoIT.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    @Test
+    public void getFieldValuesForNoFieldsReturnsEmpty() {
+        String experimentAccession = jdbcUtils.fetchRandomSingleCellExperimentAccession();
+
+        assertThat(subject.getQueryResultForMultiValueFields(experimentAccession, Optional.empty())).isEmpty();
+    }
+
+    private Iterable<String> experimentsWithMetadataProvider() {
```
Comment (alfonsomunozpomer) [inline sc/src/test/java/uk/ac/ebi/atlas/metadata/CellMetadataDaoIT.java]: In parameterized tests [method sources can return `Stream<T>`](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests-sources-MethodSource), saving you the `collect(...)` step. If you want to return unique values use [`distinct()`](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#distinct--).

File: sc/src/test/java/uk/ac/ebi/atlas/tsne/TSnePlotServiceDaoIT.java [hunk 5]
```diff
@@ -43,7 +45,7 @@ void testExpression() {
     }
 
     @Test
-    void testClusters() {
+    void testClustersForK() {
```
Comment (alfonsomunozpomer) [inline sc/src/test/java/uk/ac/ebi/atlas/tsne/TSnePlotServiceDaoIT.java]: For a few seconds I was puzzled wondering what “fork” this referred to! 😆

File: sc/src/test/java/uk/ac/ebi/atlas/tsne/TSnePlotServiceDaoIT.java [hunk 6]
```diff
@@ -56,6 +58,18 @@ void testClusters() {
...
+    @Test
+    void testClustersForPerplexity() {
+        String experimentAccession = jdbcTestUtils.fetchRandomSingleCellExperimentAccession();
+        int perplexity = jdbcTestUtils.fetchRandomPerplexityFromExperimentTSne(experimentAccession);
+
+        assertThat(subject.fetchTSnePlotForPerplexity(experimentAccession, perplexity))
+                .isNotEmpty()
+                .doesNotHaveDuplicates()
+                .extracting("name")
+                .isSubsetOf(fetchCellIds(experimentAccession));
```
Comment (alfonsomunozpomer) [inline sc/src/test/java/uk/ac/ebi/atlas/tsne/TSnePlotServiceDaoIT.java]: I ❤️  fluent APIs!

File: base/src/main/java/uk/ac/ebi/atlas/solr/cloud/fullanalytics/SingleCellAnalyticsCollectionProxy.java (hunk 2 above)
Comment (monicajianu) [reply base/src/main/java/uk/ac/ebi/atlas/solr/cloud/fullanalytics/SingleCellAnalyticsCollectionProxy.java]: Yeah, I think a lot of these methods can be refactored. They're all doing subtly different things in different context and I always forget which one is doing what.

**Pull request merged**